src/coord/timestamp.rs

```diff
@@ -1621,11 +1688,14 @@ impl Timestamper {
                                 .fetch_watermarks(&kc.topic, p, Duration::from_secs(1));
                         match watermark {
                             Ok((_low, high)) => {
-                                *current_p_offset = determine_next_offset(
-                                    *current_p_offset,
-                                    high,
+                                // high here corresponds to the next available Kafka offset.
+                                // Ex: a stream with 0 records will return a high of 0
+                                // a stream with one record (written at offset 0) will return a high of 1
+                                *current_p_offset = MzOffset::from(determine_next_offset(
+                                    Into::<KafkaOffset>::into(*current_p_offset),
+                                    KafkaOffset { offset: high - 1 },
```

Comment:
we weren't subtracting one when passing to `determine_next_offset` before, was it wrong before? Or am I missing something?

Either way, if this is indeed correct, I think it'd be better to wrap `high` in an `MzOffset` and then use the from/into impls to convert it, rather than doing manual conversions like this.

Reply:
We weren't doing the conversion into Kafka Offsets so there was no need for a -1. I was trying to be as explicit as possible in this function, and hence made the maths explicit.
Specifically: the highest Kafka timestamp that we know of is high-1. (which becomes MzOffset(high - 1 + 1), this is why there was no -1 before)

I'm happy to switch everything to be in MzOffset but I want to make sure that the -1 stays there to make the function clearer.

Reply:
I added an explicit mention of what high-1 corresponds to.

Reply:
Isn't the whole point of MzOffset that it can represent empty streams? If so - I don't understand why we're using KafkaOffset here but with -1 to represent an empty stream.

Reply:
We're not using -1 to show an empty stream. Determine_next_offset returns the maximum *current* offset that we can timestamp. This is equal to (high-1) as high returns the next available offset.

Reply:
I'm referring to this comment:

```
/// Ex: last processed offset is -1 (ak, no records have been timestamped yet)
/// The current max kafka offset is -1 (ak, the stream is empty). The function will return
/// -1. 
```

Reply:
Would you prefer determine_next_timestamp talk exclusively in MzOffsets to avoid that confusion for you?